pennylane/ops/op_math/prod.py

```diff
@@ -418,3 +430,149 @@ def _swappable_ops(op1, op2, wire_map: dict = None) -> bool:
     if np.intersect1d(wires1, wires2).size != 0:
         return False
     return np.min(wires1) > np.min(wires2)
+
+
+class ProductFactorsGrouping:
+    """Utils class used for grouping identical product factors."""
+
+    _pauli_mult = {
+        "II": (1.0, "I"),
+        "IX": (1.0, "X"),
+        "IY": (1.0, "Y"),
+        "IZ": (1.0, "Z"),
+        "XI": (1.0, "X"),
+        "XX": (1.0, "I"),
+        "XY": (1.0j, "Z"),
+        "XZ": (-1.0j, "Y"),
+        "YI": (1.0, "Y"),
+        "YX": (-1.0j, "Z"),
+        "YY": (1.0, "I"),
+        "YZ": (1.0j, "X"),
+        "ZI": (1.0, "Z"),
+        "ZX": (1.0j, "Y"),
+        "ZY": (-1.0j, "X"),
+        "ZZ": (1.0, "I"),
+    }
+    _paulis = {"X": PauliX, "Y": PauliY, "Z": PauliZ}
+
+    def __init__(self):
+        self._pauli_factors = {}  #  {wire: (pauli_coeff, pauli_word)}
+        self._non_pauli_factors = {}  # {wires: [hash, pow_coeff, operator]}
+        self.queue = ()
```

Comment:
For grouping of product factors these 3 variables are used:
- `_pauli_factors` is a dictionary that contains the necessary information to compare Pauli operators and group them if necessary.
- `_non_pauli_factors` is a dictionary that contains the necessary information to compare non-Pauli operators and group them if necessary.
- `queue` is a tuple that contains all the grouped (and immutable) factors.